service/controller/v14/resource/lifecycle/create.go

```diff
@@ -130,73 +130,67 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "start inspection of drainer configs for the guest cluster")
 
 		n := v1.NamespaceAll
 		o := metav1.ListOptions{
 			LabelSelector: fmt.Sprintf("%s=%s", key.ClusterIDLabel, key.ClusterID(customObject)),
 		}
 
-		nodeConfigs, err := r.g8sClient.CoreV1alpha1().NodeConfigs(n).List(o)
+		drainerConfigs, err := r.g8sClient.CoreV1alpha1().DrainerConfigs(n).List(o)
 		if err != nil {
 			return microerror.Mask(err)
 		}
 
-		if len(nodeConfigs.Items) > 0 {
-			for _, nodeConfig := range nodeConfigs.Items {
-				r.logger.LogCtx(ctx, "level", "debug", "message", "inspecting node config for the guest cluster")
-
-				if !nodeConfig.Status.HasFinalCondition() {
-					r.logger.LogCtx(ctx, "level", "debug", "message", "node config of guest cluster has no final state")
-					continue
-				}
+		if len(drainerConfigs.Items) > 0 {
+			for _, drainerConfig := range drainerConfigs.Items {
+				r.logger.LogCtx(ctx, "level", "debug", "message", "inspecting drainer config for the guest cluster")
 
-				r.logger.LogCtx(ctx, "level", "debug", "message", "node config of guest cluster has final state")
+				if drainerConfig.Status.HasDrainedCondition() {
+					r.logger.LogCtx(ctx, "level", "debug", "message", "drainer config of guest cluster has drained condition")
 
-				// This is a special thing for AWS. We use annotations to transport EC2
-				// instance IDs. Otherwise the lookups of all necessary information
-				// again would be quite a ball ache. Se we take the shortcut leveraging
-				// the k8s API.
-				instanceID, err := instanceIDFromAnnotations(nodeConfig.GetAnnotations())
-				if err != nil {
-					return microerror.Mask(err)
+					err := r.finishDraining(ctx, drainerConfig, workerASGName)
+					if err != nil {
+						return microerror.Mask(err)
+					}
 				}
 
-				err = r.completeLifecycleHook(ctx, instanceID, workerASGName)
-				if err != nil {
-					return microerror.Mask(err)
-				}
+				if drainerConfig.Status.HasTimeoutCondition() {
+					r.logger.LogCtx(ctx, "level", "debug", "message", "drainer config of guest cluster has timeout condition")
 
-				err = r.deleteNodeConfig(ctx, nodeConfig)
-				if err != nil {
-					return microerror.Mask(err)
+					err := r.finishDraining(ctx, drainerConfig, workerASGName)
+					if err != nil {
+						return microerror.Mask(err)
+					}
 				}
 
-				r.logger.LogCtx(ctx, "level", "debug", "message", "inspected node config for the guest cluster")
+				r.logger.LogCtx(ctx, "level", "debug", "message", "inspected drainer config for the guest cluster")
 			}
 		} else {
-			r.logger.LogCtx(ctx, "level", "debug", "message", "no node configs to inspect for the guest cluster")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "no drainer configs to inspect for the guest cluster")
 		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finished inspection of drainer configs for the guest cluster")
```

Comment:
HasDrainedCondition & HasTimeoutCondition are only useful for logging purposes correct?

Reply:
They are useful to detect a certain state. I found it complicated to find out how to move on and how to stop. Maybe the primitives should change to ease that a bit. Uncertain at the moment.